File: programmers/2021 카카오 채용연계형 인턴십/표편집.py
def solution(n, k, cmds):
    answer = ''
    nodes = {0: [n-1,1]}
    for i in range(1, n):
        if i == n-1:
            nodes[i] = [i-1, 0]
        else:
            nodes[i] = [i-1, i+1]
    stack = []
    for cmd in cmds:
        if len(cmd) > 1:
            c, x = cmd.split(' ')
            cnt = 0
            if c == 'D':
                while cnt < int(x):
                    k = nodes[k][1]
                    cnt += 1
            else:
                while cnt < int(x):
                    k = nodes[k][0]
                    cnt += 1
        else:
            if cmd == 'C':
                nodes[nodes[k][0]][1] = nodes[k][1]
                nodes[nodes[k][1]][0] = nodes[k][0]
                stack.append([k, nodes[k]])
                tmp = nodes[k]
                del nodes[k]
                if tmp[1] == 0:
                    k = tmp[0]
                else:
                    k = tmp[1]
            else:
                curr_node, val = stack.pop()
                prev_node, next_node = val
                nodes[curr_node] = [prev_node, next_node]
                nodes[prev_node][1] = curr_node
                nodes[next_node][0] = curr_node
    result = ''
    for i in range(n):
        if nodes.get(i) is None:
            result += 'X'
        else:
            result += 'O'


    return result

File: programmers/2021 카카오 채용연계형 인턴십/표편집.js
class Node {
  constructor(value, prev, next) {
    this.value = value
    this.prev = prev
    this.next = next
  }
}

function solution(n, k, cmd) {
  const table = Array.from({ length: n }, () => true)
  const nodes = Array.from({ length: n }, () => new Node())
  const removed = []
  let cur = nodes[k]
  for (let i = 0; i < n; i++) {
    nodes[i].value = i
    nodes[i].prev = nodes[i - 1]
    nodes[i].next = nodes[i + 1]
  }
  console.log(nodes)
  cmd.forEach((command) => {
    command = command.split(" ")
    switch (command[0]) {
      case "U":
        for (let i = 0; i < command[i]; i++) {
          cur = cur.prev
        }
        break
      case "D":
        for (let i = 0; i < command[i]; i++) {
          cur = cur.next
        }
        break
      case "C":
        table[cur.value] = false
        removed.push(cur)
        if (cur.next === undefined) {
          cur = cur.prev
          cur.next = undefined
        } else if (cur.prev === undefined) {
          cur.next.prev = undefined
          cur = cur.next
        } else {
          cur.prev.next = cur.next
          cur.next.prev = cur.prev
          cur = cur.next
        }
        break
      case "Z":
        const restored = removed.pop()
        table[restored.value] = true
        if (restored.prev !== undefined) {
          restored.prev.next = restored
        }
        if (restored.next !== undefined) {
          restored.next.prev = restored
        }
        break
    }
  })
  return table.map((row) => (row ? "O" : "X")).join("")
}

solution(8, 2, ["D 2", "C", "U 3", "C", "D 4", "C", "U 2", "Z", "Z"])
